spacy/pipeline/_parser_internals/nonproj.pyx

```diff
@@ -46,48 +48,76 @@ def contains_cycle(heads):
 
 def is_nonproj_arc(tokenid, heads):
     cdef vector[int] c_heads = _heads_to_c(heads)
-    return _is_nonproj_arc(tokenid, c_heads)
+    is_nonproj = _is_nonproj_arc(tokenid, c_heads)
+    if is_nonproj.second:
+        raise ValueError(Errors.E1038.format(heads=heads))
+    return is_nonproj.first
 
 
-cdef bool _is_nonproj_arc(int tokenid, const vector[int]& heads) nogil:
+cdef pair[bool, bool] _is_nonproj_arc(int tokenid, const vector[int]& heads) nogil:
```

Comment:
I think we could use C-enumerations (or even C++11 `enum class` for additional type-safety, if Cython supports it) in place of `pair[bool, bool]`? That would be more readable/explicit.

Reply:
Unless the semantics of the combinations doesn't allow mutually-exclusivity.

Reply:
I am still not sure if this is a net improvement over returning errors as values and having a template class for errors. At least, something like

```
cdef Result[int, string] myfunction(...) nogil
```

is harder to mis than an `except +` that is only visible in some `pxd`.

Reply:
That's very weakly typed. The benefit of the `Result` approach is that the error could be anything strongly-typed (like an enum class or string).